pkg/unique-id/module.go

```diff
@@ -0,0 +1,126 @@
+package unique_id
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/applike/gosoline/pkg/cfg"
+	"github.com/applike/gosoline/pkg/kernel"
+	"github.com/applike/gosoline/pkg/log"
+)
+
+type uniqueIdGenerator struct {
+	kernel.EssentialBackgroundModule
+	kernel.EssentialStage
+
+	logger     log.Logger
+	registerer InstanceRegisterer
+	generator  Generator
+	instance   Instance
+	settings   Settings
+}
+
+var um = struct {
+	sync.Mutex
+	instance *uniqueIdGenerator
+}{}
+
+func ProvideGenerator(config cfg.Config, logger log.Logger) (Generator, error) {
+	if um.instance != nil {
+		return um.instance, nil
+	}
+
+	um.Lock()
+	defer um.Unlock()
+
+	var err error
+	if um.instance, err = NewGenerator(config, logger); err != nil {
+		return nil, err
+	}
+
+	return um.instance, nil
+}
+
+func ProvideModule() kernel.ModuleFactory {
+	return func(_ context.Context, config cfg.Config, logger log.Logger) (kernel.Module, error) {
+		if um.instance != nil {
```

Comment:
Can't this be replaced with a call to `ProvideGenerator`?

Reply:
Only partially, as `Generator` and `kernel.Module` don't share any interfaces. I extracted it into a separate function